File: legacy/python/spc_up/services/report/pendencias.py
"""Pendencias CSV report for UF/exercicio (spec section 7.3)."""

from __future__ import annotations

import csv
from pathlib import Path

from sqlalchemy import or_, select
from sqlalchemy.orm import Session, selectinload

from spc_up.models.entities import Movimentacao, MovimentacaoStatus
from spc_up.services.confidence import REQUIRED_SPCA_FIELDS

_EXPORTABLE_STATUSES = (
    MovimentacaoStatus.CONFIRMADO.value,
    MovimentacaoStatus.EXPORTADO.value,
)

CSV_COLUMNS = (
    "data",
    "valor",
    "descricao",
    "motivo",
    "campos_xsd_faltantes",
    "arquivo_origem",
)


def _missing_spca_fields(movimentacao: Movimentacao) -> list[str]:
    spca = movimentacao.spca
    if spca is None:
        return list(REQUIRED_SPCA_FIELDS)

    return [
        field
        for field in REQUIRED_SPCA_FIELDS
        if not getattr(spca, field, None)
    ]


def _pendencia_motivo(movimentacao: Movimentacao, missing_fields: list[str]) -> str:
    reasons: list[str] = []
    if movimentacao.status not in _EXPORTABLE_STATUSES:
        reasons.append(f"status={movimentacao.status}")
    if movimentacao.bloqueio_export and missing_fields:
        reasons.append("campos_xsd_obrigatorios_faltantes")
    elif movimentacao.bloqueio_export:
        reasons.append("bloqueio_export")
    if movimentacao.confianca_global < 0.60:
        reasons.append("confianca_baixa")
    return "; ".join(reasons) if reasons else "pendencia"


def generate_pendencias_csv(
    session: Session,
    uf: str,
    exercicio: int,
    output_path: str | Path,
) -> int:
    """Write pendencias CSV and return the number of rows written."""
    movimentacoes = session.execute(
        select(Movimentacao)
        .options(
            selectinload(Movimentacao.spca),
            selectinload(Movimentacao.arquivo_ingestao),
        )
        .where(
            Movimentacao.uf == uf,
            Movimentacao.exercicio == exercicio,
            or_(
                Movimentacao.status.not_in(_EXPORTABLE_STATUSES),
                Movimentacao.bloqueio_export.is_(True),
            ),
        )
        .order_by(Movimentacao.data_movimento, Movimentacao.id)
    ).scalars().all()

    path = Path(output_path)
    path.parent.mkdir(parents=True, exist_ok=True)

    with path.open("w", newline="", encoding="utf-8") as handle:
        writer = csv.DictWriter(handle, fieldnames=CSV_COLUMNS)
        writer.writeheader()
        for movimentacao in movimentacoes:
            missing = _missing_spca_fields(movimentacao)
            arquivo_origem = ""
            if movimentacao.arquivo_ingestao is not None:
                arquivo_origem = movimentacao.arquivo_ingestao.nome_arquivo

            writer.writerow(
                {
                    "data": movimentacao.data_movimento.isoformat(),
                    "valor": f"{movimentacao.valor:.2f}",
                    "descricao": movimentacao.descricao_raw,
                    "motivo": _pendencia_motivo(movimentacao, missing),
                    "campos_xsd_faltantes": ",".join(missing),
                    "arquivo_origem": arquivo_origem,
                }
            )

    return len(movimentacoes)

File: packages/core/src/report/pendencias.ts
/** Pendencias CSV report for UF/exercicio (spec section 7.3). */

import { createWriteStream } from "node:fs";
import { mkdir } from "node:fs/promises";
import { dirname } from "node:path";

import { and, eq, notInArray, or } from "drizzle-orm";

import {
  arquivoIngestao,
  movimentacao,
  movimentacaoSpca,
  type Db,
} from "@spc-up/db";

import { REQUIRED_SPCA_FIELDS } from "../confidence";

const EXPORTABLE_STATUSES = ["CONFIRMADO", "EXPORTADO"] as const;

export const CSV_COLUMNS = [
  "data",
  "valor",
  "descricao",
  "motivo",
  "campos_xsd_faltantes",
  "arquivo_origem",
] as const;

type SpcaRow = typeof movimentacaoSpca.$inferSelect;

function missingSpcaFields(spca: SpcaRow | null | undefined): string[] {
  if (spca == null) {
    return [...REQUIRED_SPCA_FIELDS];
  }

  const fieldMap: Record<(typeof REQUIRED_SPCA_FIELDS)[number], keyof SpcaRow> = {
    fonte_recurso: "fonteRecurso",
    natureza_recurso: "naturezaRecurso",
    tipo_origem_recurso: "tipoOrigemRecurso",
  };

  return REQUIRED_SPCA_FIELDS.filter((field) => {
    const key = fieldMap[field];
    const value = spca[key];
    return value == null || value === "";
  });
}

function pendenciaMotivo(
  status: string,
  bloqueioExport: boolean,
  confiancaGlobal: number,
  missingFields: string[],
): string {
  const reasons: string[] = [];
  if (!EXPORTABLE_STATUSES.includes(status as (typeof EXPORTABLE_STATUSES)[number])) {
    reasons.push(`status=${status}`);
  }
  if (bloqueioExport && missingFields.length > 0) {
    reasons.push("campos_xsd_obrigatorios_faltantes");
  } else if (bloqueioExport) {
    reasons.push("bloqueio_export");
  }
  if (confiancaGlobal < 0.6) {
    reasons.push("confianca_baixa");
  }
  return reasons.length > 0 ? reasons.join("; ") : "pendencia";
}

function escapeCsv(value: string): string {
  if (/[",\n\r]/.test(value)) {
    return `"${value.replace(/"/g, '""')}"`;
  }
  return value;
}

/** Write pendencias CSV and return the number of rows written. */
export async function generatePendenciasCsv(
  db: Db,
  uf: string,
  exercicio: number,
  outputPath: string,
): Promise<number> {
  const ufUpper = uf.toUpperCase();

  const rows = await db
    .select({
      dataMovimento: movimentacao.dataMovimento,
      valor: movimentacao.valor,
      descricaoRaw: movimentacao.descricaoRaw,
      status: movimentacao.status,
      bloqueioExport: movimentacao.bloqueioExport,
      confiancaGlobal: movimentacao.confiancaGlobal,
      spca: movimentacaoSpca,
      nomeArquivo: arquivoIngestao.nomeArquivo,
    })
    .from(movimentacao)
    .leftJoin(movimentacaoSpca, eq(movimentacaoSpca.movimentacaoId, movimentacao.id))
    .leftJoin(arquivoIngestao, eq(arquivoIngestao.id, movimentacao.arquivoIngestaoId))
    .where(
      and(
        eq(movimentacao.uf, ufUpper),
        eq(movimentacao.exercicio, exercicio),
        or(
          notInArray(movimentacao.status, [...EXPORTABLE_STATUSES]),
          eq(movimentacao.bloqueioExport, true),
        ),
      ),
    )
    .orderBy(movimentacao.dataMovimento, movimentacao.id);

  await mkdir(dirname(outputPath), { recursive: true });

  await new Promise<void>((resolve, reject) => {
    const stream = createWriteStream(outputPath, { encoding: "utf-8" });
    stream.on("error", reject);
    stream.on("finish", () => resolve());

    stream.write(`${CSV_COLUMNS.join(",")}\n`);

    for (const row of rows) {
      const missing = missingSpcaFields(row.spca);
      const line = [
        row.dataMovimento,
        Number(row.valor).toFixed(2),
        escapeCsv(row.descricaoRaw),
        escapeCsv(
          pendenciaMotivo(
            row.status,
            row.bloqueioExport,
            row.confiancaGlobal,
            missing,
          ),
        ),
        escapeCsv(missing.join(",")),
        escapeCsv(row.nomeArquivo ?? ""),
      ].join(",");
      stream.write(`${line}\n`);
    }

    stream.end();
  });

  return rows.length;
}
